seaborn/axisgrid.py

```diff
@@ -1797,6 +1797,47 @@ def plot_marginals(self, func, **kwargs):
 
         return self
 
+    def refline(self, *, x=None, y=None, joint=True, marginal=True, **line_kws):
+        """Add a reference line to joint and/or marginal axes.
+
+        Parameters
+        ----------
+        x, y : numeric
+            Value to draw the line at.
+        joint, marginal : bools
+            Whether to add the reference line to the joint/marginal axes.
+        line_kws : key, value mappings
+            Other keyword arguments are passed to :meth:`matplotlib.axes.Axes.axvline`
+            when ``x`` is not None or :meth:`matplotlib.axes.Axes.axhline` when ``y``
+            is not None.
+
+        Returns
+        -------
+        :class:`JointGrid` instance
+            Returns ``self`` for easy method chaining.
+
+        """
+        if x is None and y is None:
+            raise ValueError('Must provide ``x`` or ``y``')
```

Comment:
I'm not sure about this check (or the one below that requires either `joint` or `marginal` to be True).

These checks will make life difficult for any sort of code that wraps `JointGrid` and exposes the refline parameters. They'll need to modify the logic of their code based on the parameter values, instead of just passing them through to seaborn.

It can be helpful to provide feedback to users when surprising behavior would otherwise happen. But I think that the API here is simple enough that calling it with no value for x or y will be trivial to debug, so it's not worth the complexity it would add to downstream code. Do you disagree?

Reply:
> These checks will make life difficult for any sort of code that wraps JointGrid and exposes the refline parameters. They'll need to modify the logic of their code based on the parameter values, instead of just passing them through to seaborn.

I don't fully understand what you mean by this. Are you trying to avoid raising any exceptions in the seaborn code for this method? What logic would they need to modify?

Regarding the check for either `joint` or `marginal` to be `True`, if both are `False` the effect of the method call would be nothing. This feels like it is failing silently.

Without the first check, the resulting error would actually be quite confusing due to the implementation (because we don't yet know if we should use `axhline()` or `axvline()`, same for which marginal axes):
```
   1833         if joint:
-> 1834             getattr(self.ax_joint, func)(value, **line_kws)
   1835         if marginal:
   1836             getattr(ax_marg, func)(value, **line_kws)

UnboundLocalError: local variable 'func' referenced before assignment
```

Of course, the implementation can be altered to first assign `func` to say `axhline()` and then allow matplotlib to raise an error when it receives `None`:
```
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-8-0de983514f78> in <module>
      5 # by default plot on both the marginal and joint axes
      6 # g.refline(linestyle='dashed', color='gray')
----> 7 g.ax_joint.axvline(None)

~/seaborn_dev/lib/python3.7/site-packages/matplotlib/axes/_axes.py in axvline(self, x, ymin, ymax, **kwargs)
    785         # Strip away the units for comparison with non-unitized bounds.
    786         xx, = self._process_unit_info([("x", x)], kwargs)
--> 787         scalex = (xx < xmin) or (xx > xmax)
    788 
    789         trans = self.get_xaxis_transform(which='grid')

TypeError: '>' not supported between instances of 'float' and 'NoneType'
```
I'm fine with changing the implementation and removing that check, but I don't think the check should be removed without changing the implementation. Do you prefer any exceptions to come from matplotlib?

Reply:
My point is, imagine you are writing a function that uses `JointGrid`. You want to expose `refx`, `refy` parameters, defaulting to `None`. With your exceptions, that function would need to add in some logic

```python
def wrapper(..., refx=None, refy=None, ...):
    g = JointGrid(...)
    ...
    if refx is not None or refy is not None:
        g.refline(refx, refy)
    ...
```
Not the most complicated code in the world, but a little annoying.

> Do you prefer any exceptions to come from matplotlib?

No, I think that calling `ax{h,v}line` should only happen when `x` or `y` are not `None`, I just think that logic should be internal to the method.

I'm not really concerned about silently doing nothing if both `x` and `y` are `None` (or with `marginal=False, joint=False`). I don't think that's an error; you're doing what the user asked for.

Reply:
Ok, that's helpful.  I was developing this for more direct use, but since we often use wrapper functions in seaborn it makes sense that we may want to bubble up this functionality in the future &ndash; in which case, removing the checks makes it easier. I'll remove the them to align with the seaborn philosophy 🙂